ext/timers/01_timers.js

```diff
@@ -61,534 +48,307 @@
 
-          // update helpers
-          if (gp !== null) {
-            ggp = gp;
-          }
-          gp = p;
-          p = node;
-          node = node.getChild(dir);
-        }
+  /**
+   * The keys in this map correspond to the key ID's in the spec's map of active
+   * timers. The values are the timeout's cancel rid.
+   *
+   * @type {Map<number, number>}
+   */
+  const activeTimers = new Map();
 
-        // update root
-        this.#root = head.right;
-      }
+  let nextId = 1;
 
-      // make root black
-      this.#root.red = false;
+  /**
+   * @param {Function | string} callback
+   * @param {number} timeout
+   * @param {Array<any>} args
+   * @param {boolean} repeat
+   * @param {number | undefined} prevId
+   * @returns {number} The timer ID
+   */
+  function initializeTimer(
+    callback,
+    timeout,
+    args,
+    repeat,
+    prevId,
+  ) {
+    // 2. If previousId was given, let id be previousId; otherwise, let
+    // previousId be an implementation-defined integer than is greater than zero
+    // and does not already exist in global's map of active timers.
+    let id;
+    let cancelRid;
+    if (prevId !== undefined) {
+      // `prevId` is only passed for follow-up calls on intervals
+      assert(repeat);
+      id = prevId;
+      cancelRid = MapPrototypeGet(activeTimers, id);
+    } else {
+      // TODO(@andreubotella): Deal with overflow.
+      // https://github.com/whatwg/html/issues/7358
+      id = nextId++;
+      cancelRid = core.opSync("op_timer_handle");
 
-      return ret;
+      // Step 4 in "run steps after a timeout".
+      MapPrototypeSet(activeTimers, id, cancelRid);
     }
 
-    /** Returns `true` if removed, `false` if not found. */
-    remove(data) {
-      if (this.#root === null) {
-        return false;
-      }
-
-      const head = new RBNode(null); // fake tree root
-      let node = head;
-      node.right = this.#root;
-      let p = null; // parent
-      let gp = null; // grand parent
-      let found = null; // found item
-      let dir = 1;
-
-      while (node.getChild(dir) !== null) {
-        const last = dir;
-
-        // update helpers
-        gp = p;
-        p = node;
-        node = node.getChild(dir);
-
-        const cmp = this.#comparator(data, node.data);
-
-        dir = cmp > 0;
+    // 3. If the surrounding agent's event loop's currently running task is a
+    // task that was created by this algorithm, then let nesting level be the
+    // task's timer nesting level. Otherwise, let nesting level be zero.
+    // 4. If timeout is less than 0, then set timeout to 0.
+    // 5. If nesting level is greater than 5, and timeout is less than 4, then
+    // set timeout to 4.
+    timeout = MathMax(timeout, (timerNestingLevel > 5) ? 4 : 0);
+
+    // 9. Let task be a task that runs the following steps:
+    const task = {
+      action: () => {
+        // 1. If id does not exist in global's map of active timers, then abort
+        // these steps.
```

Comment:
This comment is a little superfluous because it's obvious _what_ the code below does, just not _why_ it does it.

It'd be more helpful to explain that (IIUC) a timer is appended to the `timerTasks` array when created but not deleted from said array when it's disarmed before expiration, hence the necessity to have this side check.